### PR TITLE
EDF: adding code sample + getter function

### PR DESCRIPTION
**Pull request by alexpaschoaletto**
This PR simply includes a sample for the Earliest Deadline First (EDF) scheduler and adds a getter function to a thread deadline - `k_thread_deadline_get()`. This prevents the application code from needing to access the thread structure directly to retrieve this information.

Signed-off-by: Alexander Paschoaletto <alexp@isep.ipp.pt>

### CODE REVIEW

File: include/zephyr/kernel.h [hunk 1]
```diff
@@ -958,6 +958,22 @@ __syscall void k_thread_priority_set(k_tid_t thread, int prio);
...
+ * scheduler when deciding which thread has the highest priority.
+ *
+ * @note You should enable @kconfig{CONFIG_SCHED_DEADLINE} in your project
+ * configuration.
+ * 
+ * @param thread Thread from which to get the deadline
+ * @return absolute deadline value, in cycle units
+ *
+ */
+int32_t k_thread_deadline_get(k_tid_t thread);
```
Comment (andyross) [inline include/zephyr/kernel.h:975]: What's the use case you're imagining here? Basically I have a hard time imagining a use case for a function like this that isn't a huge race conditions factory.  If you just need it for the sample, you can just as easily compute it (subject to the same raciness!) from the time of the original deadline call, no?

Also: the returned value is an absolute cycle time e.g. as you might see from k_cycle_get_32(), and not a delta from current time as passed to k_thread_deadline_set().  That asymmetry is a footgun and should be called out in the docs.

**Review by andyross — COMMENTED** [review]
Sample looks clean to me.  Not really loving the API, which seems a little dangerous.

File: include/zephyr/kernel.h (hunk 1 above)
Comment (alexpaschoaletto) [reply include/zephyr/kernel.h:975]: I'm sorry, why would this cause a race condition? It's just a getter, and deadlines are not expected to change so often anyway.

The use case is... if the application code need to know what deadline value was set. If this function does not exist, the application will have to access `thread->base.prio_deadline` and that is not good practice AFAIK or even consistent with the API anyway. Just like `k_thread_name_get` is a wrapper for `thread->name` and `k_thread_priority_get` is a wrapper for `thread->priority`.

**Comment by kartben** [comment]
> So... anyone else to review / add to the discussion above? 😅

@alexpaschoaletto If you could get the CI back on track (conflict that needs resolving, compliance check error due to having to squash commits together, ....) I think that will help attract reviewers come back to this. Thanks in advance!

**Comment by alexpaschoaletto** [comment]
> @alexpaschoaletto If you could get the CI back on track (conflict that needs resolving, compliance check error due to having to squash commits together, ....) I think that will help attract reviewers come back to this. Thanks in advance!

Thank you for the tip, I will look into those and report back when they get fixed. I might need help with CI corrections here and there, but I'll raise my hand when needed.

**Comment by alexpaschoaletto** [comment]
Well, I think it's all there. Commits are properly divided into kernel and sample, and all checks seem to be passing. `k_reschedule` contribution from #83249 is applied as well.

**Comment by andyross** [comment]
Swinging back by...

> The use case is... if the application code need to know what deadline value was set. 

Yeah, that's not a use case. :)  A use case is a story explaining **why** application code needs to know what deadline was set.

Basically: app code is responsible for deciding how long a thread's tasks will take and expressing that as a deadline.  If you need to know what you did, then clearly you did it wrong.  If you need to recompute it, you don't need to know what it used to be. 

Basically this is an API with a bad smell.  I can't think of a usage that doesn't seem like a bug workaround.   Not a -1 from me (though not a +1 either) as it's simple and harmless.  But I'm dubious about the value.
